docs/developer-docs/smart-contracts/advanced-features/http-gateways.mdx

```diff
@@ -0,0 +1,105 @@
+
+# HTTP gateways
+
+<MarkdownChipRow labels={["Advanced", "Tutorial"]} />
+
+## Overview
+
+The ICP HTTP gateway protocol enables conventional HTTP clients to interact with the ICP network. Through the HTTP gateway, software such as web browsers can retrieve and digest canister code and assets, such as HTML, JavaScript, images, or videos. The HTTP gateway makes this workflow possible by translating standard HTTP requests into API canister calls that ICP can understand.
+
+There are several HTTP gateway nodes running on ICP, but it should be noted that an HTTP gateway node could become compromised, creating a threat for canisters. Compromised gateways are threats even for canisters that are considered 'secure', such that they use certify assets, use a default client, and do not maintain their own certification checking.
+
+For a more secure solution, it is possible to run your own HTTP gateway instance locally.
+
+### Terminology
+
+- **Request**: A message sent to a server or endpoint.
+
+- **Client**: A service that is able to send an HTTP formatted request and receive a response from a server.
+
+- **Gateway**: Enables the transfer of inbound and outbound messages.
+
+## HTTP gateway workflow
+
+On ICP, an HTTP request is handled by an HTTP client using the following workflow:
+
+1. An HTTP client makes an outbound request.
+
+2. The HTTP gateway intercepts the request and resolves the canister ID of the request's destination.
+
+3. The request is encoded with Candid and sent in a query call to the destination canister's `http_request` method.
+
+4. ICP's API gateway intercepts this request, resolves the canister ID based on the hostname address, and then forwards the request to the relevant replica node. All requests made to ICP through an HTTP gateway are forwarded through this API gateway.
+
+5. The canister receives and processes the request, then returns the response.
+
+6. The HTTP gateway decodes the Candid encoding on the response and sends the request again via an update call to the canister's `http_request_update` method. The update call goes through consensus on the subnet.
+
+7. If necessary, the HTTP gateway fetches additional response body data through query calls.
+
+8. The HTTP gateway validates the response's certificate, if applicable.
+
+9. The HTTP gateway returns the decoded response to the HTTP client.
+
+Additional details can be found in the [HTTP gateway specification](https://internetcomputer.org/docs/current/references/http-gateway-protocol-spec).
+
+## Running a local HTTP gateway
+
+To run your own local HTTP gateway, a proof-of-concept implementation can be used that enables a secure end-to-end connection with canisters deployed on ICP.
+
+This implementation features:
+
+- Translation between HTTP asset requests and ICP API calls.
+
+- Detects ICP domains from principals and custom domain records.
+
+- Terminates TLS connections locally.
+
+- Bypasses remote gateway denylists.
```

Comment:
Do we really want to advertise this?

Reply:
We also advertise it on the [http-proxy](https://github.com/dfinity/http-proxy/) docs.

I think it's important for the community to know. There've been more than a few instances of someone complaining about the boundary node deny lists on the forum. I think they should know that the answer is to run their own, local gateway.